### PR TITLE
LOOKDEVX-146 : Add in runtime file I/O for look information 

### PR DESCRIPTION
**Pull request by bernardkwok**
LOOKDEVX-4, LOOKDEVX-146: 
- Add ability to read/write collections, materialassigns, looks, and look groups
- Add in geom support for collections.
- Update unit tests.
- Add in version option on read to allow for material-node conversion (by default) on read.

### CODE REVIEW

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 1]
```diff
@@ -371,6 +373,158 @@ namespace
...
+    PvtPrim* readCollection(const CollectionPtr& src, PvtPrim* parent, PvtStage* stage)
+    {
+        const RtToken name(src->getName());
+
+        PvtPrim* prim = stage->createPrim(parent->getPath(), name, RtCollection::typeName());
+        RtCollection collection(prim->hnd());
+        collection.getIncludeGeom().setValueString(src->getIncludeGeom());
+        collection.getExcludeGeom().setValueString(src->getExcludeGeom());
+
+        // createCollectionConnections() handles creating collection inclusion
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Is this comment is misplaced?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 2]
```diff
@@ -371,6 +373,158 @@ namespace
...
+        // connections.
+        return prim;
+    }
+
+    void createCollectionConnections(const vector<CollectionPtr>& collectionElements, PvtPrim* parent)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: The name of this function seems misleading to me, but maybe that's just me.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 3]
```diff
@@ -371,6 +373,158 @@ namespace
...
+            rassign.getCollection().addTarget(collection->hnd());
+
+            PvtPrim* material = findPrimOrThrow(RtToken(assign->getMaterial()), parent);
+            rassign.getMaterial().addTarget(material->hnd());
+
+            rassign.getExclusive().getValue().asBool() = assign->getExclusive();
+            rassign.getGeom().getValue().asString() = assign->getActiveGeom();
+        }
+
+        // createLookConnections() handles creating look inheritance
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Also a misplaced comment?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 4]
```diff
@@ -371,6 +373,158 @@ namespace
...
+
+        PvtPrim* rootPrim = stage->getRootPrim();
+
+        // Read collections
+        for (const ElementPtr& elem : doc->getCollections())
+        {
+            if (!filter || filter(elem))
+            {
+                // Make sure the element has not been loaded already.
+                PvtPath path("/" + elem->getName());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Is there some constant for the "/" you can use?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 5]
```diff
@@ -371,6 +373,158 @@ namespace
...
+            }
+        }
+
+        // Read looks
+        for (const LookPtr& elem : doc->getLooks())
+        {
+            if (!filter || filter(elem))
+            {
+                // Make sure the element has not been loaded already.
+                PvtPath path("/" + elem->getName());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Same "/" comment.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 1 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp]: It's supposed to indicate that this functionality is not done here but in the order function.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 6]
```diff
@@ -371,6 +373,158 @@ namespace
...
+            }
+        }
+
+        // Read look groups
+        for (const LookGroupPtr& elem : doc->getLookGroups())
+        {
+            if (!filter || filter(elem))
+            {
+                // Make sure the element has not been loaded already.
+                PvtPath path("/" + elem->getName());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Same "/" comment.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 2 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp]: Up for suggestions. It's making connections between collections that have "include" relationships.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 3 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp]: Similar to the previous one. It's a note that connection part of the work is not done here.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 7]
```diff
@@ -700,6 +860,108 @@ namespace
         }
     }
 
+    void writeCollections(PvtStage* stage, DocumentPtr dest, RtWriteOptions::WriteFilter filter)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: Should this be DocumentPtr& to avoid the increment?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 8]
```diff
@@ -371,6 +373,158 @@ namespace
...
+        const RtToken name(src->getName());
+
+        PvtPrim* prim = stage->createPrim(parent->getPath(), name, RtLook::typeName());
+        RtLook look(prim->hnd());
+
+        // Create material assignments
+        for (const MaterialAssignPtr assign : src->getMaterialAssigns())
+        {
+            PvtPrim* aprim = stage->createPrim(parent->getPath(), RtToken(assign->getName()), RtMaterialAssign::typeName());
+            RtMaterialAssign rassign(aprim->hnd());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: for consistency, should this be rtAssign?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 9]
```diff
@@ -700,6 +860,108 @@ namespace
...
+                collection->setIncludeGeom(rtCollection.getIncludeGeom().getValueString());
+
+                RtRelationship rtIncludeCollection = rtCollection.getIncludeCollection();
+                string includeList = rtIncludeCollection.getTargetsAsString();                
+                collection->setIncludeCollectionString(includeList);
+            }
+        }
+    }
+
+    void writeLooks(PvtStage* stage, DocumentPtr dest, RtWriteOptions::WriteFilter filter)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: DocumentPtr& ?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 10]
```diff
@@ -700,6 +860,108 @@ namespace
...
+                            break;
+                        }
+                        massign->setGeom(rtMatAssign.getGeom().getValueString());
+                    }
+                }
+            }
+        }
+    }
+
+    void writeLookGroups(PvtStage* stage, DocumentPtr dest, RtWriteOptions::WriteFilter filter)
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp]: DocumentPtr& ?

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 1 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp]: I see

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 2 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp]: You seem to use this create...Connections pattern throughout, not a big deal I guess.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 8 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp]: or better yet you use rtMatAssign later on.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 7 above)
Comment (JGamache-autodesk) [reply source/MaterialXRuntime/RtFileIo.cpp]: The C++ guidelines even go further: use a Document& in context where a function does not modify ownership.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-smartptrparam

Comment (feldstj) [reply source/MaterialXRuntime/RtFileIo.cpp]: That makes sense actually.

**Pull request merged**